From 9ab5bc7f93cad86c13d922c406038cf07dcfdfbf Mon Sep 17 00:00:00 2001
From: Ziad Elhanafy <ziad.elhanafy@arm.com>
Date: Tue, 19 Sep 2023 09:14:39 +0100
Subject: [PATCH] RSS: Make hardcoded MHUv3 driver values configurable

Multiple values such as the shared outband buffer base, the shared outband
buffer size and the flag value used for doorbell communication are hardcoded.
Make them configurable by adding them the MHUv3 device structure.

Upstream-Status: Pending
Signed-off-by: Ziad Elhanafy <ziad.elhanafy@arm.com>
---
 .../arm/rss/common/native_drivers/mhu_v3_x.h  | 11 +++++--
 .../common/native_drivers/mhu_wrapper_v3_x.c  | 30 +++++++++++--------
 .../kronos/device/host_device_definition.c    |  6 ++++
 3 files changed, 33 insertions(+), 14 deletions(-)

diff --git a/platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.h b/platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.h
index dc884ff8a..158663b44 100644
--- a/platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.h
+++ b/platform/ext/target/arm/rss/common/native_drivers/mhu_v3_x.h
@@ -23,6 +23,7 @@
 #define __MHU_V3_X_H__
 
 #include <stdint.h>
+#include <stddef.h>
 #include <stdbool.h>
 #include "host_base_address.h"
 
@@ -59,11 +60,11 @@ extern "C" {
 #define MHU3_PBX_DBCH_FLAG_AP_COMMS     UINT32_C(0x2)
 
 /* Size of shared outband message buffer in bytes */
-#define MHU3_OUTBAND_BUF_SIZE UINT64_C(0x2000)
+#define MHU3_AP_RSS_OUTBAND_BUF_SIZE UINT64_C(0x2000)
 #define MHU3_OUTBAND_BUF_HEADER_SIZE UINT64_C(0x4)
 
 /* Base of shared outband mesage buffer */
-#define MHU3_OUTBAND_BUF_BASE HOST_AP_RSS_MAILBOX_BASE_S
+#define MHU3_AP_RSS_OUTBAND_BUF_BASE HOST_AP_RSS_MAILBOX_BASE_S
 
 /**
  * \brief MHUv3 error enumeration types.
@@ -105,6 +106,12 @@ struct mhu_v3_x_dev_t {
     const uintptr_t base;
     /* Type of the MHUv3 frame */
     enum mhu_v3_x_frame_t frame;
+    /* Base of shared outband message buffer */
+    const uintptr_t outband_buf_base;
+    /* Size of shared outband message buffer in bytes */
+    const size_t outband_buf_size;
+    /* PBX_DBCH_FLAG */
+    const uint32_t pbx_dbch_flag;
     /* Minor revision of the MHUv3 */
     uint32_t subversion;
     /* Flag to indicate if the MHUv3 is initialized */
diff --git a/platform/ext/target/arm/rss/common/native_drivers/mhu_wrapper_v3_x.c b/platform/ext/target/arm/rss/common/native_drivers/mhu_wrapper_v3_x.c
index e8df55ca5..80d389d72 100644
--- a/platform/ext/target/arm/rss/common/native_drivers/mhu_wrapper_v3_x.c
+++ b/platform/ext/target/arm/rss/common/native_drivers/mhu_wrapper_v3_x.c
@@ -141,19 +141,23 @@ fail:
     return error_mapping_to_mhu_error_t(err);
 }
 
-size_t mhu_get_max_message_size(void)
+size_t mhu_get_max_message_size(void *mhu_dev)
 {
+    struct mhu_v3_x_dev_t *dev = mhu_dev;
+
     /*
      * First 4 bytes represent the MHU msg header. Rest of buffer
      * holds the actual message.
      */
-    return (MHU3_OUTBAND_BUF_SIZE - MHU3_OUTBAND_BUF_HEADER_SIZE);
+    return (dev->outband_buf_size - MHU3_OUTBAND_BUF_HEADER_SIZE);
 }
 
-static int validate_buffer_params(uintptr_t buf_addr, size_t buf_size)
+static int validate_buffer_params(void *mhu_dev,
+                                  uintptr_t buf_addr,
+                                  size_t buf_size)
 {
     if ((buf_addr == 0) || (!IS_ALIGNED(buf_addr, 4)) ||
-            buf_size > mhu_get_max_message_size()) {
+            buf_size > mhu_get_max_message_size(mhu_dev)) {
         return MHU_V_3_X_ERR_INVALID_PARAM;
     }
 
@@ -227,7 +231,7 @@ enum mhu_error_t mhu_send_data(void *mhu_sender_dev,
         return error_mapping_to_mhu_error_t(MHU_V_3_X_ERR_NONE);
     }
 
-    err = validate_buffer_params((uintptr_t)send_buffer, size);
+    err = validate_buffer_params(mhu_sender_dev, (uintptr_t)send_buffer, size);
     if (err != MHU_V_3_X_ERR_NONE) {
         goto fail;
     }
@@ -236,11 +240,11 @@ enum mhu_error_t mhu_send_data(void *mhu_sender_dev,
      * First 4 bytes represents the header which currently stores the size
      * of msg. Rest of buffer holds the actual message.
      */
-    (void)memcpy((void *)MHU3_OUTBAND_BUF_BASE, (void *)&msg_len,
+    (void)memcpy((void *)dev->outband_buf_base, (void *)&msg_len,
             MHU3_OUTBAND_BUF_HEADER_SIZE);
 
     /* Copy the message */
-    (void)memcpy((void *)(MHU3_OUTBAND_BUF_BASE + MHU3_OUTBAND_BUF_HEADER_SIZE),
+    (void)memcpy((void *)(dev->outband_buf_base + MHU3_OUTBAND_BUF_HEADER_SIZE),
             (void *)send_buffer, msg_len);
     /*
      * Use only one channel 0 to ring the doorbell on receiver for
@@ -250,7 +254,7 @@ enum mhu_error_t mhu_send_data(void *mhu_sender_dev,
      * bits on the same channel or else transmitter will continue waiting
      * on the status bit indefinitely.
      */
-    err = signal_and_wait_for_clear(dev, 0, MHU3_PBX_DBCH_FLAG_AP_COMMS);
+    err = signal_and_wait_for_clear(dev, 0, dev->pbx_dbch_flag);
     if (err != MHU_V_3_X_ERR_NONE)
         goto fail;
 
@@ -268,7 +272,9 @@ enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
     uint32_t msg_len;
     enum mhu_v3_x_error_t err;
 
-    err = validate_buffer_params((uintptr_t)receive_buffer, *size);
+    err = validate_buffer_params(mhu_receiver_dev,
+                                 (uintptr_t)receive_buffer,
+                                 *size);
     if (err != MHU_V_3_X_ERR_NONE) {
         goto fail;
     }
@@ -278,7 +284,7 @@ enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
      * attestation report and measurements, while the actual data is shared
      * through a common memory region in Secure/Root PAS.
      */
-    err = wait_for_signal_and_clear(dev, 0, MHU3_PBX_DBCH_FLAG_AP_COMMS);
+    err = wait_for_signal_and_clear(dev, 0, dev->pbx_dbch_flag);
     if (err != MHU_V_3_X_ERR_NONE) {
         goto fail;
     }
@@ -287,7 +293,7 @@ enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
      * First 4 bytes represents the header which currently stores the size
      * of msg. Rest of buffer holds the actual message.
      */
-    (void)memcpy((void *)&msg_len, (void *)MHU3_OUTBAND_BUF_BASE,
+    (void)memcpy((void *)&msg_len, (void *)dev->outband_buf_base,
             MHU3_OUTBAND_BUF_HEADER_SIZE);
     if (*size < msg_len) {
         return MHU_V_3_X_ERR_UNSUPPORTED;
@@ -296,7 +302,7 @@ enum mhu_error_t mhu_receive_data(void *mhu_receiver_dev,
     *size = msg_len;
 
     /* Copy the message */
-    (void)memcpy((void *)receive_buffer, (void *)(MHU3_OUTBAND_BUF_BASE +
+    (void)memcpy((void *)receive_buffer, (void *)(dev->outband_buf_base +
                 MHU3_OUTBAND_BUF_HEADER_SIZE), msg_len);
 
     return MHU_V_3_X_ERR_NONE;
diff --git a/platform/ext/target/arm/rss/kronos/device/host_device_definition.c b/platform/ext/target/arm/rss/kronos/device/host_device_definition.c
index 95ba4eba8..d421118ed 100644
--- a/platform/ext/target/arm/rss/kronos/device/host_device_definition.c
+++ b/platform/ext/target/arm/rss/kronos/device/host_device_definition.c
@@ -129,6 +129,9 @@ struct atu_dev_info SCP_ATU_INFO = {
 struct mhu_v3_x_dev_t MHU_AP_TO_RSS_DEV = {
     .base = MHU2_RECEIVER_BASE_S,
     .frame = MHU_V3_X_MBX_FRAME,
+    .outband_buf_base = MHU3_AP_RSS_OUTBAND_BUF_BASE,
+    .outband_buf_size = MHU3_AP_RSS_OUTBAND_BUF_SIZE,
+    .pbx_dbch_flag = MHU3_PBX_DBCH_FLAG_AP_COMMS,
     .subversion = 0};
 #endif
 
@@ -136,6 +139,9 @@ struct mhu_v3_x_dev_t MHU_AP_TO_RSS_DEV = {
 struct mhu_v3_x_dev_t MHU_RSS_TO_AP_DEV = {
     .base = MHU2_SENDER_BASE_S,
     .frame = MHU_V3_X_PBX_FRAME,
+    .outband_buf_base = MHU3_AP_RSS_OUTBAND_BUF_BASE,
+    .outband_buf_size = MHU3_AP_RSS_OUTBAND_BUF_SIZE,
+    .pbx_dbch_flag = MHU3_PBX_DBCH_FLAG_AP_COMMS,
     .subversion = 0};
 #endif
 
-- 
2.25.1

